### PR TITLE
Store: Don't error when no stores are matched

### PR DESCRIPTION
**Pull request by SuperQ**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

It's normal and not an error if a query does not match due to no downstream stores. This is common when querying with external labels and tiered query servers.

This bug was introduced in https://github.com/thanos-io/thanos/pull/5296

Fixes: https://github.com/thanos-io/thanos/issues/5862

### CODE REVIEW

**Comment by SuperQ** [comment]
:eyeglasses: @bwplotka

**Comment by SuperQ** [comment]
I also made a small optimization to avoid allocating the error message every time.

File: pkg/store/proxy.go [hunk 1]
```diff
@@ -35,6 +35,10 @@ type ctxKey int
 // StoreMatcherKey is the context key for the store's allow list.
 const StoreMatcherKey = ctxKey(0)
 
+// ErrorNoStoresMatched is returned if the query does match any data.
```
Comment (yeya24) [inline pkg/store/proxy.go:38]: does not match

**Review by yeya24 — APPROVED** [review]

File: pkg/store/proxy.go (hunk 1 above)
Comment (SuperQ) [reply pkg/store/proxy.go]: Fixed.

**Review by yeya24 — APPROVED** [review]
Thanks

**Comment by fpetkovski** [comment]
@GiedriusS since the k-way merge optimization added this functionality, any concerns with merging this PR?

**Review by bwplotka — APPROVED** [review]
Not great fan of this as people with misconfigured single querier can be confused by no results.. but it might be consistent and valid response for your query - and it's blocking some users with chained Queries. 

Valid trade-off, so let's go! Thanks @SuperQ

**Comment by bwplotka** [comment]
I don't think @GiedriusS did it on purpose - it was just side effect.

**Pull request merged**